### PR TITLE
Implement a simple container for wamp messages

### PR DESCRIPTION
**Pull request by davidchappelle**
By providing a simple container for wamp messages:

- Easy to move towards a model where we can support different
  types of serialization over the transport (i.e. JSON).
- Makes mangement of msgpack zones simpler.
- Provides a common data type that can be used for abstracting
  out wamp transports.
- Offers a potential base class for creating a heirarchy of
  message types.

### CODE REVIEW

**Comment by davidchappelle** [comment]
@oberstet This is phase one of merging in the work to abstract out wamp transports. The next phase will include the transport abstraction.

**Comment by oberstet** [comment]
I like that!

> For in process transports, this allows us to actually pass the message ownership off directly to the application router, which helps to avoid any unnecessary copying of data.

This is nice.

---

I'd like to merge, but it seems to have conflicts - maybe because I merged https://github.com/crossbario/autobahn-cpp/pull/87

**Comment by davidchappelle** [comment]
Ok. I'll rebase from master again and fix the new abort message handling to use the new message definition.

**Comment by oberstet** [comment]
@davidchappelle awesome! thanks a lot.

**Comment by davidchappelle** [comment]
@oberstet Ready to merge now.

File: autobahn/wamp_message.hpp [hunk 1]
```diff
@@ -31,13 +31,150 @@
...
+/*!
+ * A class that represents a wamp message in its simplest form.
+ *
+ *     [field1, field2, field3, ...]
+ *
+ * Although not all arrays are valid wamp messages, this class
+ * simply provides the building blocks to construct any type of
+ * message.
+ *
+ * TODO: Investigate the benefits of creating a heirarchy of
```
Comment (jpetso) [inline autobahn/wamp_message.hpp]: Spelling: `hierarchy`

File: autobahn/wamp_session.ipp [hunk 2]
```diff
@@ -231,71 +234,56 @@ boost::future<uint64_t> wamp_session<IStream, OStream>::join(
...
-    m_io.dispatch([=]() {
+    m_io_service.dispatch([=]() {
         auto shared_self = weak_self.lock();
         if (!shared_self) {
             return;
         }
 
         if (m_session_id) {
-            throw protocol_error("session already joined");
+            m_session_join.set_exception(protocol_error("session already joined"));
```
Comment (jpetso) [inline autobahn/wamp_session.ipp]: It seems you can add a line "Fixes #90" to the commit message, and it will auto-close issue #90. Likely also #85.

**Comment by jpetso** [comment]
Nice work Dave :)

**Comment by oberstet** [comment]
Awesome! `wamp_message` allows abstraction from serialization format and underlying transport. Big step forward for AutobahnCpp. Dave, thanks a lot for your work!

**Pull request merged**